### DISCUSSION TITLE
Accept SPDX compliant license specifications?

### DISCUSSION THREAD

**Discussion by cxreiff** [Homebrew/discussions#5561] [Tap maintenance and Homebrew development]
### Output of `brew config`

```shell
N/A
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

I am a Rust developer, setting up a workflow using [cargo-dist](https://github.com/axodotdev/cargo-dist) to automatically commit formulae to my personal homebrew tap. `cargo-dist` just copies the standard SPDX license string specified in the `Cargo.toml` file (e.g. `MIT OR Apache-2.0`) and uses it for the license in the formula; however this fails the `brew test-bot --only-tap-syntax`/`brew audit` test, as Homebrew has its own special syntax for licenses ([as mentioned here](https://docs.brew.sh/License-Guidelines)).

```
==> brew audit --except=installed --tap=cxreiff/tap
==> FAILED
Full audit --except=installed --tap=cxreiff/tap output
  cxreiff/tap/ttysvr
    * Formula ttysvr contains non-standard SPDX licenses: ["MIT OR Apache-2.0"].
```

Couple of questions:

1. It surprised me a bit that the "standard" license format was invalid—is there a specific reason for it? Would it make sense for Homebrew itself to support the standard format, or would I have better luck writing a PR for `cargo-dist` with some sort of conversion layer?
2. Is anybody aware of existing code for doing this kind of conversion, either within the repo or provided by some third party?

Thank you!

**Comment by Bo98** [comment]
I don't recall the precise historical reasons for (1) but it's unlikely to change unless you can argue strongly for it. I think the original idea was it was avoided the need to do string parsing (since we would be converting it to an array internally anyway) and made more sense in a language like Ruby where arrays can be used directly. There was no conversion functions at all originally.

For (2) there is a conversion here: https://github.com/Homebrew/brew/blob/277c4259fb5326fa13dc8d132b986be21c333b97/Library/Homebrew/utils/spdx.rb#L127-L159. It's private API but the logic shouldn't change should you wish to port it. It's fairly new in comparison. Licenses in formulae have been around since 2020 but this code was only added in 2023 for the specific scenario of parsing the formulae.brew.sh API which did use original SPDX syntax (as that API was originally only intended for outside users). Perhaps something we'll tidy up down the line if we make separate external and internal formulae.brew.sh API.

**Reply by cxreiff** [reply]
Oh great! Exactly what I was looking for, sounds like the way forward is porting that to Rust and contributing to cargo-dist. Thank you for your help!

**Reply by cxreiff** [reply]
Regarding (1), not pushing hard for it as I personally have a work-around here, but my strongest arguments for Homebrew supporting it would be:

- The Arch Linux PKGBUILD format is [transitioning to SPDX strings](https://rfc.archlinux.page/0016-spdx-license-identifiers/).
- Crates.io and NPM already use it.
- Releasing to multiple package managers is common and a plain SPDX string is a language-agnostic and package-manager-agnostic representation that already has adoption.
- Therefore Homebrew supporting the plain string would mean that without any fancy parsing, someone could release their project to Crates.io, NPM, the Arch User Repository, AND Homebrew.

**Reply by cxreiff** [reply]
Counter argument, while the syntax is different Homebrew already uses SPDX _identifiers_, and the conversion is not wildly complicated.
